modules/dlm/src/main/java/org/elasticsearch/dlm/DataLifecycleService.java

```diff
@@ -277,15 +342,29 @@ static TimeValue getRetentionConfiguration(DataStream dataStream) {
     }
 
     /**
-     * Calculate the age of the index since creation or rollover time if the index was already rolled.
-     * The rollover target is the data stream name the index is a part of.
+     * Action listener that records the encountered failure using the provided recordError callback for the
+     * provided target index. If the listener is notified of success it will clear the recorded entry for the provided
+     * target index using the clearErrorRecord callback.
      */
-    static TimeValue getCreationOrRolloverDate(String rolloverTarget, IndexMetadata index) {
-        RolloverInfo rolloverInfo = index.getRolloverInfos().get(rolloverTarget);
-        if (rolloverInfo != null) {
-            return TimeValue.timeValueMillis(rolloverInfo.getTime());
-        } else {
-            return TimeValue.timeValueMillis(index.getCreationDate());
+    static class ErrorRecordingActionListener implements ActionListener<Void> {
```

Comment:
This doesn't actually record or clear any errors, it only calls some `Consumer` methods. Perhaps it would be better to pass in `DataLifecycleErrorStore` and make this class opinionated about calling `errorStore.clear(...)` and `errorStore.record(targetIndex, e)`?

Reply:
++ great idea